### PR TITLE
Add PeekRandom

### PR DESCRIPTION
**Pull request by tonyhb**
This adds PeekRandom to the queue, enabling us to peek from random offsets if we encounter purely key queue issues.

## Type of change (choose one)
- [x] Chore (refactors, upgrades, etc.)
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Security fix (non-breaking change that fixes a potential vulnerability)
- [ ] Docs
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)

## Checklist
- [ ] I've linked any associated issues to this PR.
- [ ] I've tested my own changes.

*[Check our Pull Request Guidelines](https://github.com/inngest/inngest/blob/main/docs/PULL_REQUEST_GUIDELINES.md)*

### CODE REVIEW

File: pkg/execution/state/redis_state/queue_processor.go [hunk 1]
```diff
@@ -780,6 +788,19 @@ func (q *queue) processPartition(ctx context.Context, p *QueuePartition, guarant
...
+		// We hit custom concurrency key issues.  Re-process this partition at a random offset, as long
+		// as random offset is currently false (so we don't loop forever)
+
+		// Note: we must requeue the partition to remove the lease.
+		err := q.PartitionRequeue(ctx, q.primaryQueueShard, p, q.clock.Now().Truncate(time.Second).Add(PartitionConcurrencyLimitRequeueExtension), true)
+		if err != nil {
+			log.From(ctx).Warn().Err(err).Msg("error requeuieng partition for random peek")
+		}
+
+		return q.processPartition(ctx, p, guaranteedCapacity, true)
```
Comment (darwin67) [inline pkg/execution/state/redis_state/queue_processor.go:801]: process after requeue?

File: pkg/execution/state/redis_state/queue.go [hunk 2]
```diff
@@ -1638,8 +1646,41 @@ func (q *queue) Peek(ctx context.Context, partition *QueuePartition, until time.
...
+	}
+	partitionKey := partition.zsetKey(q.primaryQueueShard.RedisClient.kg)
+	return q.peek(
+		ctx,
+		q.primaryQueueShard,
+		peekOpts{
+			Limit:        limit,
+			Until:        until,
+			PartitionKey: partitionKey,
+			Random:       true,
```
Comment (darwin67) [inline pkg/execution/state/redis_state/queue.go:1676]: could this be made an configuration option?
so it's possible to enable it selectively without impacting everyone at once.

File: pkg/execution/state/redis_state/queue_processor.go (hunk 1 above)
Comment (tonyhb) [reply pkg/execution/state/redis_state/queue_processor.go:801]: to force a random peek offset

File: pkg/execution/state/redis_state/lua/queue/peek.lua [hunk 3]
```diff
@@ -7,10 +7,24 @@ Peek returns items from the queue that are unleased and the vesting time <= peek
...
+local limit        = tonumber(ARGV[2])
+local randomOffset = ARGV[3]
 
-local itemIds = redis.call("ZRANGE", queueIndex, "-inf", peekUntil, "BYSCORE", "LIMIT", 0, limit)
+
+local offset = 0
+
+if randomOffset == "1" then
+	local count = redis.call("ZRANGE", queueIndex, "-inf", peekUntil)
+	if count > limit and sequential == 0 then
```
Comment (darwin67) [inline pkg/execution/state/redis_state/lua/queue/peek.lua:19]: where's `sequential` coming from?

**Review by darwin67 — APPROVED** [review]

**Pull request merged**